codex/validation.nim

```diff
@@ -117,16 +132,41 @@ proc run(validation: Validation) {.async.} =
     trace "Validation stopped"
     discard
   except CatchableError as e:
-    error "Validation failed", msg = e.msg
+    error "Validation failed", msg = e.msg, groups = validation.config.groups,
+      groupIndex = validation.config.groupIndex
+
+proc epochForDurationBackFromNow(validation: Validation,
+    duration: times.Duration): SecondsSince1970 =
+  return validation.clock.now - duration.inSeconds
+
+proc restoreHistoricalState(validation: Validation) {.async} =
+  logScope:
+    groups = validation.config.groups
+    groupIndex = validation.config.groupIndex
+  trace "Restoring historical state..."
+  let startTimeEpoch = validation.epochForDurationBackFromNow(MaxStorageRequestDuration)
+  let slotFilledEvents = await validation.market.queryPastSlotFilledEvents(
+    fromTime = startTimeEpoch)
+  trace "Found slot filled events", numberOfSlots = slotFilledEvents.len
+  for event in slotFilledEvents:
+    let slotId = slotId(event.requestId, event.slotIndex)
+    if validation.shouldValidateSlot(slotId):
+      trace "Adding slot [historical]", slotId
+      validation.slots.incl(slotId)
+  trace "Removing slots that have ended..."
+  await removeSlotsThatHaveEnded(validation)
+  trace "Historical state restored", numberOfSlots = validation.slots.len
 
 proc start*(validation: Validation) {.async.} =
   validation.periodicity = await validation.market.periodicity()
   validation.proofTimeout = await validation.market.proofTimeout()
   await validation.subscribeSlotFilled()
+  await validation.restoreHistoricalState()
   validation.running = validation.run()
```

Comment:
Yes, great catch, I noticed this as well while reviewing 🙏 🙌  It's on my list of items to fix in nim-codex.

Reply:
what is the problem with this call? We do not seem to discard the future result and we do await on it later in `stop`. As I can read in https://status-im.github.io/nim-chronos/tips.html#discard, using `asyncSpawn` would make exception handling more aggressive (any unhandled exception would cause a defect and crash), but it does not return anything, so looks like you cannot wait on it later in order to stop gracefully. Trying to learn here...

Reply:
I also see we have our own utility `asyncSpawn` in `codex/utils/asyncspawn.nim`.

Reply:
Thanks for extra insights. I also discussed this with Mark. We do have to recored the `Future` here so that we can use it later when we cancel in `stop`, but to make sure we do not swallow exceptions, we need call it with `asyncSpawn` and make sure that what we do leak any exceptions because that would cause the deadly defect. BTW: I know that @emizzle is working on aligning `asyncSpawn` in other places, so I guess it is best if I do not add it myself here, or at least if probably better to have it in a small separate PR...

Reply:
That is correct, futures need to be `await`ed or `asyncSpawn`ed, else exceptions in async routines will get silently swallowed. This is especially import in async loops that can silently fail. I am currently working on these changes. No need to handle them.